Adding a method in `vector.py`

MohitKumar020291 opened the pull request:
<!-- Your title above should be a short description of what
was changed. Do not include the issue number in the title. -->

#### References to other Issues or PRs
<!-- If this pull request fixes an issue, write "Fixes #NNNN" in that exact
format, e.g. "Fixes #1234" (see
https://tinyurl.com/auto-closing for more information). Also, please
write a comment on that issue linking back to this pull request once it is
open. -->
https://github.com/sympy/sympy/issues/25915


#### Brief description of what is fixed or changed


#### Other comments


#### Release Notes

<!-- Write the release notes for this release below between the BEGIN and END
statements. The basic format is a bulleted list with the name of the subpackage
and the release note for this PR. For example:

* solvers
  * Added a new solver for logarithmic equations.

* functions
  * Fixed a bug with log of integers. Formerly, `log(-x)` incorrectly gave `-log(x)`.

* physics.units
  * Corrected a semantical error in the conversion between volt and statvolt which
    reported the volt as being larger than the statvolt.

or if no release note(s) should be included use:

NO ENTRY

See https://github.com/sympy/sympy/wiki/Writing-Release-Notes for more
information on how to write release notes. The bot will check your release
notes automatically to see if they are formatted correctly. -->

<!-- BEGIN RELEASE NOTES -->
* vector
  * Added a method equals to vector class
<!-- END RELEASE NOTES -->

MohitKumar020291 committed:
Adding an if case in expr.py

oscarbenjamin commented on sympy/core/expr.py:
```diff
@@ -843,6 +843,12 @@ def equals(self, other, failing_expression=False):
         if constant not in (True, None) and constant != 0:
             return False
+        
+        from sympy import radsimp
+        self = radsimp(simplify(self))
+        other = radsimp(other)
+        if self == other:
+            return True
 
```
I don't think that this really addresses the underlying problem. I made some suggestions here: https://github.com/sympy/sympy/issues/25915#issuecomment-1824604943

MohitKumar020291 commented:
Thanks Oscar for your feedback! I checked and repeated the code in the discussion of issue, and I found that:
1.) `diff = factor_terms(simplify(A*Q.i - 2*Q.i), radical=True)`, as diff is zero, it represents `True`. Is it because of the boolean system of sympy?
2.) Now as `diff` is `True`, then `equals` will return `False` (according to `equals` method's code, `not diff` condition in if block), which is incorrect, because if `diff` is `0` then it should return `True`.
3.) I think that the `equals` method is sufficient in itself, and the main issue is due to this `if` block about `diff` and may be the confusion of zero as `True` or `False`. Or it is done for some reason? (as I am new to this repo, not too much sure.)

I am attaching a screenshots to make my point more clear and equals method code:
![image](https://github.com/sympy/sympy/assets/120097428/7a0129ef-2324-4c30-ab3e-206c4c6df8d3)
--------------------------------------------------------------------------------------------------------------------
![image](https://github.com/sympy/sympy/assets/120097428/7d4dd267-3f45-445f-a1d5-a4e93ec96f0f)

oscarbenjamin commented:
The problem is that the test in equals is for `is_zero` but that only applies to numbers and not to vectors.

Probably the best fix here is for vectors to have a `.equals` method that extracts the numeric scalar components and calls `.equals` on them.

MohitKumar020291 committed:
Added new method to vector_equals()

MohitKumar020291 commented:
Thanks for your feedback! I have added a new method `vector_equals`, which extracts the scaler components and this is also useful for vectors with their directions, like i, j, k.

oscarbenjamin commented:
Rather than adding a `vector_equals` method to `Expr` what is eneded is to add an`.equals` method to the vector types that can override the `Expr.equals` method.

I don't actually know what is a good way to extract the scalar components from a vector.

@moorepants if we want to implement something like `v1.equals(v2)` what is a good way to extract the vector components of the two vectors. I suppose that they might not be in the same coordinate system but we would want to extract their components in a way that we can do `simplify(v1i - v2i)` so we do not want the expressions to be too complicated.

MohitKumar020291 commented:
Before making new changes, it would be nice if you let me know whether you agree with this logic or not, thanks!

oscarbenjamin commented:
It should be a vector.equals method that extracts the scalar components and calls .equals on them.

MohitKumar020291 commented:
I am not sure that whether I should directly add this method to `sympy/vector/vector.py`'s `class Vector` or should make a new class which inherit from `class Vector`? And I have modified the code which now also uses the `class Expr`'s `equals` method as per feedback. Thanks!

moorepants commented:
> @moorepants if we want to implement something like v1.equals(v2) what is a good way to extract the vector components of the two vectors. I suppose that they might not be in the same coordinate system but we would want to extract their components in a way that we can do simplify(v1i - v2i) so we do not want the expressions to be too complicated.

You would have to first express the two vectors in the same reference frame and then extract the scalars for comparison.

MohitKumar020291 commented:
> > @moorepants if we want to implement something like v1.equals(v2) what is a good way to extract the vector components of the two vectors. I suppose that they might not be in the same coordinate system but we would want to extract their components in a way that we can do simplify(v1i - v2i) so we do not want the expressions to be too complicated.
> 
> You would have to first express the two vectors in the same reference frame and then extract the scalars for comparison.

Can you please confirm, like the way I did is correct?

MohitKumar020291 committed:
Added a class CustomVector in vector.py

MohitKumar020291 committed:
Added a class CustomVector in vector.py

MohitKumar020291 committed:
Added a class CustomVector in vector.py

MohitKumar020291 committed:
Added a method equals in class Vector

MohitKumar020291 commented:
I have added a method `equals` to class Vector and also added a test in `sympy\core\tests\test_expr.py` to check whether it is working fine or not.
Here is the SS:
![image](https://github.com/sympy/sympy/assets/120097428/ded4f926-018e-4dcd-9bdd-426d2ff4817b)

MohitKumar020291 committed:
Added a method equals in class Vector

oscarbenjamin commented:
The tests are not running in CI because there is a merge conflict.

MohitKumar020291 commented:
How should I proceed?

oscarbenjamin commented:
The merge conflict needs to be resolved.

MohitKumar020291 committed:
Updating the branch

MohitKumar020291 committed:
updated

MohitKumar020291 committed:
solved the main resolve conflicts

MohitKumar020291 committed:
Blank commit line-266

MohitKumar020291 committed:
test

MohitKumar020291 committed:
test

MohitKumar020291 committed:
test

MohitKumar020291 committed:
test

oscarbenjamin commented:
You can run `bin/test quality` before pushing rather than pushing and waiting for CI checks to run.

MohitKumar020291 committed:
test

MohitKumar020291 committed:
test

MohitKumar020291 committed:
Vector.zero fix

MohitKumar020291 committed:
test_functions.py

MohitKumar020291 committed:
test_functions.py

MohitKumar020291 committed:
test_functions.py

MohitKumar020291 committed:
test_functions.py

MohitKumar020291 committed:
test_functions.py

MohitKumar020291 committed:
author: add Mohit Kumar to .mailmap

MohitKumar020291 commented:
Resolved conflicts.

oscarbenjamin commented:
Okay, so conflicts are resoled and tests are passing which is good.

Now, as mentioned previously what the Vector.equals method should be doing is just extracting the vector components and calling .equals on them.

As @moorepants said the two vectors would first need to be converted to a common coordinate system or at least the method should check that they are in the same coordinate system before doing anything else.

I don't know how to convert the vectors to a common coordinate system so you will need to read the vector docs to find that out.

MohitKumar020291 committed:
modified vector.py

MohitKumar020291 committed:
modified vector.py

MohitKumar020291 committed:
modified vector.py

MohitKumar020291 committed:
modified vector.py

MohitKumar020291 committed:
modified vector.py

MohitKumar020291 committed:
modified vector.py

MohitKumar020291 commented:
I have implemented the suggestions made.

oscarbenjamin commented on sympy/vector/vector.py:
```diff
@@ -344,6 +344,70 @@ def _div_helper(one, other):
+                    if type(i) != int and type(j) != int:
+                        result.append(i.equals(j))
+                    else:
+                        result.append(radsimp(simplify(S_S[index])) == radsimp(simplify(S_O[index])))
```
Why are `radsimp` and `simplify` being used here?

I deally this should only call `equals`.

MohitKumar020291 replied:
because for scaler components, which could be int (probably always), are not supported by equals of expr.

oscarbenjamin commented on sympy/vector/vector.py:
```diff
@@ -344,6 +344,70 @@ def _div_helper(one, other):
+        from sympy import simplify, radsimp
+        from sympy.vector import Vector
+
+        # finding vector components of self and other with their respective coordinate system
+        V_S = [0, 0, 0]
+        V_O = [0, 0, 0]
+
+        # finding scalar components of self and other
+        S_S = [0, 0, 0]
+        S_O = [0, 0, 0]
+        if self == Vector.zero:
+            V_S = [0, 0, 0]
+            S_S = [0, 0, 0]
+        else:
+            index = 0
+            for i in self.components:
+                if isinstance(i, sympy.vector.vector.BaseVector):
+                    V_S[index] = i
+                    S_S[index] = self.components[i]
+                    index += 1
+
+        if other == Vector.zero:
+            V_O = [0, 0, 0]
+            S_O = [0, 0, 0]
+        else:
+            index = 0
+            for i in other.components:
+                if isinstance(i, sympy.vector.vector.BaseVector):
+                    V_O[index] = i
+                    S_O[index] = other.components[i]
+                index += 1
```
I don't know if this code is correct but I am sure that there is a better way to express both vectors in the same coordinate system. There should be an existing function for doing this.

MohitKumar020291 replied:
there is a function used like, `CoordSys3D('self')`, but it has some drawbacks for the purpose in PR/issue/tests written, like:
It is only to check whether vectors are in same coordinate system or not. With some limits:
  * Q = CoordSys3D('Q')
    i, j, k = Q.base_vectors()
    A = sympy.sqrt(2) + sympy.sqrt(6)) / (sympy.sqrt(sympy.sqrt(3) + 2))
    v1 = 2*i +  2*j
    v2 =  A*i + 2*j
    #when we do use `CoordSys3D`, it will transform like:
    v1 = 2*Q.i + 2*Q.j
    v2 = ((sqrt(2) + sqrt(6))/sqrt(sqrt(3) + 2))*Q.i + 2*Q.j

    Can we compare them without using components for coordinate systems? No. 

    Why? As the expressions are not in simplified form, and we cannot perform `radsimp` on `CoordSys3D`

    Can we use equals method on them? No. CoordSys3D do not support equals.

    Why not using `components` on CoordSys3D and compare keys and values using equals? Again, CoordSys3D do 
    not support `components`.

    Why not using `args`? Example if v1 = Q.i, then v1.args returns a tuple (0, Q), and when for loop used on it scaler 
    components comes to be `0` and vector comes to be coordinate system not proper vector (which happens when 
    there is a scaler component of a vector except 1). And for Vector.zero, it returns an empty tuple, again not so much 
    helpful.

So, the best way is using loops on `components`, to find the scaler and vector components. And if a vector is zero (like Vector.zero), simply returning a list of [0, 0, 0]. 


With your experience, if you have a better method to do it, please let me know. Thanks.

oscarbenjamin replied:
I don't have much experience with the vector module so I don't know what the right way to do this is but it is definitely not the code that is here. There are existing functions for expressing a vector in different coordinate systems and those should be used.

oscarbenjamin commented on sympy/vector/vector.py:
```diff
@@ -344,6 +344,70 @@ def _div_helper(one, other):
+            else:
+                return False
+        except Exception as e:
+            raise RuntimeError(f"An error occurred: {str(e)}") from e
```
There is no point in catching the exception to turn it into a RuntimeError like this. The original exception should be allowed to propagate.

MohitKumar020291 committed:
Original exception

oscarbenjamin commented on sympy/vector/vector.py:
```diff
@@ -344,6 +344,70 @@ def _div_helper(one, other):
+            else:
+                return False
+        except Exception as e:
+            raise e
```
There is no point catching the exception just to raise it again.

MohitKumar020291 committed:
resolved

MohitKumar020291 committed:
resolved

MohitKumar020291 committed:
resolved

oscarbenjamin commented:
The function that should be used is express:
https://docs.sympy.org/latest/modules/vector/coordsys.html#expression-of-quantities-in-different-coordinate-systems
```
In [11]: from sympy.vector import CoordSys3D, express
    ...: from sympy.abc import a, b, c
    ...: N = CoordSys3D('N')
    ...: M = N.orient_new_axis('M', a, N.k)

In [12]: v = 2*N.i+z*N.k

In [13]: v
Out[13]: (2) i_N + (z) k_N

In [14]: express(v, M)
Out[14]: (2⋅cos(a)) i_M + (-2⋅sin(a)) j_M + (z) k_M
```

MohitKumar020291 commented:
> The function that should be used is express: https://docs.sympy.org/latest/modules/vector/coordsys.html#expression-of-quantities-in-different-coordinate-systems
> 
> ```
> In [11]: from sympy.vector import CoordSys3D, express
>     ...: from sympy.abc import a, b, c
>     ...: N = CoordSys3D('N')
>     ...: M = N.orient_new_axis('M', a, N.k)
> 
> In [12]: v = 2*N.i+z*N.k
> 
> In [13]: v
> Out[13]: (2) i_N + (z) k_N
> 
> In [14]: express(v, M)
> Out[14]: (2⋅cos(a)) i_M + (-2⋅sin(a)) j_M + (z) k_M
> ```

In my new code, I have implemented the approach that checks whether they are in same coordinate system or not, using `Expr.equals`, but I do not think that we should use `express` as `components` extracts the vector components which are `expressed` their respective coordinate system. And they are also sent to basic `Expr.equals` in V_C, so to check that whether they are in same coordinate system or not.

oscarbenjamin commented:
It is not enough to check if they are in the same coordinate system. Both vectors need to be expressed in a single coordinate system or otherwise `v1.equals(v2)` might return False even though the vectors are equal.

MohitKumar020291 committed:
used express

MohitKumar020291 committed:
used express

MohitKumar020291 commented:
Implemented use of `express`.

oscarbenjamin commented on sympy/vector/vector.py:
```diff
@@ -344,6 +344,56 @@ def _div_helper(one, other):
+            self_coord = list((self.components).keys())[0].args[1]
+            other_coord = list((other.components).keys())[0].args[1]
+
+            # expressing self in coordinate system, N.
+            N = self_coord.locate_new('N', self)
+            self = express(N.position_wrt(self_coord), N)
+
+            N = other_coord.locate_new('N', other)
+            other = express(N.position_wrt(other_coord), N)
```
This does not look like it is expressing both vectors in a single coordinate system.

MohitKumar020291 replied:
They are in same (or single) coordinate system.

oscarbenjamin commented:
I don't think that the logic here is correct. The`.equals` method can return True, False or None using three-valued logic. All three cases need to be handled correctly:
https://docs.sympy.org/latest/guides/booleans.html

There should be more tests like tests for vectors that should not be equal and tests for vectors that are in different coordinate systems.

MohitKumar020291 committed:
used three-valued logic

MohitKumar020291 commented:
> There should be more tests like tests for vectors that should not be equal and tests for vectors that are in different coordinate systems.

Yes, this approach is included in V_C which checks whether they are in same coordinate system or not, as finally we returns V_C `and` S_C. Is it what you meant or anything else just like adding more functionality to `Vector` , for example, `Vector.scaler_equals` which will check whether their scaler components are equal or not?
